### PR TITLE
util,store/tikv: trace caller of cancel() by a new WithCancel

### PR DESCRIPTION
**Pull request by tiancaiamao**
When canceled error occurs, it's difficult to debug the cause of error.
This PR replace standard `WithCancel` by a customized one, to provide file:line information
when cancel function is called.

PTAL @hhkbp2

### CODE REVIEW

File: util/context_cancel_debugger.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import (
+	"fmt"
+	"runtime"
+	"time"
+
+	"github.com/juju/errors"
+	"golang.org/x/net/context"
+)
+
+var debug = false
```
Comment (hhkbp2) [inline util/context_cancel_debugger.go]: Enable by default?

**Comment by hhkbp2** [comment]
Rest LGTM
Nice design without change to the interface.

File: util/context_cancel_debugger.go (hunk 1 above)
Comment (tiancaiamao) [inline util/context_cancel_debugger.go]: Enable be default will have some performance cost, I prefer disable by default. @hhkbp2 
Or should its value got from environment  variable?  @shenli

**Comment by tiancaiamao** [comment]
PTAL @shenli @zimulala @hhkbp2

**Comment by shenli** [comment]
LGTM
Please use this branch to test running sysbench on tidb cluster and kill one tikv.

**Comment by hhkbp2** [comment]
LGTM

**Comment by shenli** [comment]
@hhkbp2 PTAL

**Pull request merged**